### ISSUE TITLE
Potential memory leak while Gradle app build benchmark

### ISSUE THREAD

**Issue by FireHook** [apollographql/apollo-kotlin#6266]
### Version

> 4.0.0

### Summary

I did the migration from **Apollo** `2.5.12` version with such dependencies:
```
apollographql-apollo-api-jvm = { module = "com.apollographql.apollo:apollo-api-jvm", version.ref = "apollo" }
apollographql-apollo-coroutines-support = { module = "com.apollographql.apollo:apollo-coroutines-support", version.ref = "apollo" }
apollographql-apollo-runtime = { module = "com.apollographql.apollo:apollo-runtime", version.ref = "apollo" }
apollographql-apollo-rx3-support = { module = "com.apollographql.apollo:apollo-rx3-support", version = "2.5.14" }
```

To **Apollo** `4.0.0` with such dependencies:
```
apollographql-apollo-runtime = { module = "com.apollographql.apollo:apollo-runtime", version.ref = "apollo" }
apollographql-apollo-normalized-cache = { module = "com.apollographql.apollo:apollo-normalized-cache", version.ref = "apollo" }
```

I have noticed that Gradle Profiler throws `java.lang.RuntimeException: Multiple Gradle daemons were used.` exception during Gradle app build benchmarking using `gradle-profiler --gradle-user-home ../benchmark --benchmark --measure-config-time --profile buildscan --scenario-file scripts/buildprofiler/clean_build_scenario --output-dir apk_debug_build`, where clean_build_scenario is five times iteration of `:app:assembleDebug` with `--no-build-cache` and  `--no-configuration-cache` arguments.

Further investigation shown that Memory leaks might be occurred. 

**Gradle Profiler Meatspace memory usage while app benchmarking with Apollo 2.5.12**
<img width="498" alt="Screenshot 2024-11-20 at 16 05 48" src="https://github.com/user-attachments/assets/7461ea23-b5b5-4ff6-aa75-6b65d51a3f9a">

**Gradle Profiler Meatspace memory usage while app benchmarking with Apollo 4.0.0**
<img width="498" alt="Screenshot 2024-11-20 at 16 05 25" src="https://github.com/user-attachments/assets/27a3db7c-8c61-40fa-a6f6-1fce7977cd7f">

As you can see from screenshots while Apollo 4.0.0 were used, the Meatspace memory used by Gradle increases within each iteration, so it looks like these are memory leaks.

I have tried updating Apollo to newer 4.1.0 version but the problem does not eliminate.

Could you, please, take a look what can cause the issue?

### Logs

```
(Your logs here)
```

**Comment by martinbonnin** [comment]
Thanks for opening this. `4.0.0` uses a separate classloader for running codegen in order to support [compiler plugins](https://www.apollographql.com/docs/kotlin/advanced/compiler-plugins). That sounds like a potential culprit. I'll dig into this.

**Comment by martinbonnin** [comment]
@FireHook I could reproduce. There is a Gradle issue about this: https://github.com/gradle/gradle/issues/18313

As a workaround, #6267 introduces `Service.useProcessIsolation` to run the required Apollo tasks in a separate process and therefore reclaim the memory and the end of the task when the process terminates.

It's a tradeoff though as the task may take a bit more time to start (forking, etc...).

In the future, I'm planning to migrate the Gradle plugin to [Gratatouille](https://github.com/GradleUp/gratatouille/). This will give us more control over the classloader. We'll need to remove `Service.packageNameGenerator` and `Service.operationOutputGenerator` first though. 

In the meantime, if the leak is too annoying for you, set `useProcessIsolation` to true:

```kotlin
apollo {
  service("service") {
    packageName.set("com.exampel")
    useProcessIsolation.set(true)
  }
}
```

**Comment by martinbonnin** [comment]
Update: we investigated this with @scana and ended with a more universal solution. Latest SNAPSHOTs do not need `useProcessIsolation.set(true)` anymore. That will be available in 4.1.1
